File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        Stack<TreeNode> sk = new Stack<>();
        TreeNode cur = root, prev = null;
        
        while (cur != null || !sk.isEmpty()) {
            while (cur != null) {
                sk.push(cur);
                cur = cur.left;
            }
            if (sk.peek().right == null || sk.peek().right == prev) {
                prev = sk.pop();
                list.add(prev.val);
            } else {
                cur = sk.peek().right;
            }
        }
        return list;
    }
}

File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        return post(root, ans);
    }
    
    vector<int> post(TreeNode* root, vector<int>& ans) {
        if (root == NULL)
            return ans;
        post(root->left, ans);
        post(root->right, ans);
        ans.push_back(root->val);
        
        return ans;
    }
};